File: Data Structures & Algorithms/maximum-product-subarray/submission-1.java
class Solution {
    public int maxProduct(int[] nums) {
        int[][] dp = new int[2][nums.length];
        dp[0][0] = nums[0];
        dp[1][0] = nums[0];
        int max = nums[0];
        
        for (int i = 1; i < nums.length; i++) {
            dp[0][i] = Math.max(nums[i], Math.max(dp[0][i - 1] * nums[i], dp[1][i - 1] * nums[i]));
            max = Math.max(max, dp[0][i]);
            dp[1][i] = Math.min(nums[i], Math.min(dp[0][i - 1] * nums[i], dp[1][i - 1] * nums[i]));
        }

        return max;
    }
}

File: Data Structures & Algorithms/max-water-container/submission-1.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        l = 0
        r = len(heights) - 1
        water = 0
        
        while l < r:
            height = min(heights[l], heights[r])
            water = max(water, height * (r - l))
            if heights[l] < heights[r]:
                l += 1
            else:
                r -= 1
        
        return water